update monitoring page (#108)
ConsenSys/doc.tessera: docs/HowTo/Use/Monitoring.md

# Monitor Tessera

> Tessera can be used with InfluxDB and Prometheus time-series databases to record API usage metrics.

Reviewer: Might be replace 'and' with 'or'. It used to be that we needed both, but now either works and prom & grafana are our standard and we're consistent

The data recorded can be visualized by using an existing dashboard tool such as Grafana.

> In addition, Tessera logs can be searched, analyzed, and monitored using Splunk.

Reviewer: Logs can be done via Splunk or ELK - we've got an example of Tessera logs in ELK in the quickstart too. Ref links: https://github.com/ConsenSys/quorum-dev-quickstart/blob/master/files/common/logstash/pipeline/20_tessera.conf and https://github.com/ConsenSys/quorum-dev-quickstart/blob/master/files/common/filebeat/filebeat.yml

You can set up Splunk such that the logs for multiple Tessera nodes in a network are accessible from a single
centralized Splunk instance.

## API metrics

Tessera can record the following usage metrics for each endpoint of its API:

* Average response time
* Maximum response time

> * Minimum Response Time

Reviewer 2: ```suggestion
* Minimum response time
```

* Request count
* Requests per second

You can store these metrics in an InfluxDB or Prometheus time-series database for further analysis.

* Use [InfluxDB](https://www.influxdata.com/time-series-platform/influxdb/) when you prefer for metrics to be "pushed"
  from Tessera to the database.
  For example, Tessera starts a service which periodically writes the latest metrics to the database by calling the
  database's API.
* Use [Prometheus](https://prometheus.io/) when you prefer for metrics to be "pulled" from Tessera by the database.
  For example, Tessera exposes a `/metrics` API endpoint which the database periodically calls to fetch the latest metrics.

Both databases integrate well with the open source dashboard editor [Grafana](https://grafana.com/) to allow for easy
creation of dashboards to visualize the data being captured from Tessera.

### Using InfluxDB

> See the [InfluxDB documentation](https://docs.influxdata.com/influxdb) for details on how to set up an InfluxDB database
> ready for use with Tessera.

Reviewer 2: ```suggestion
The [InfluxDB documentation](https://docs.influxdata.com/influxdb) provides details on how to set up an InfluxDB database
for use with Tessera.
```

    ```

!!! info

    You can call the InfluxDB HTTP API directly as an alternative to using the `influx` CLI.

You can configure each Tessera server type (for example, `P2P`, `Q2T`, `ADMIN`, `THIRDPARTY`, `ENCLAVE`) to store API
metrics in an InfluxDB.
You can configure these servers to store metrics to the same database or separate ones.

> Not all servers need to be configured to store metrics.

Reviewer 2: Existing issue, but I'm not sure what this means. Does it mean that we can optionally configure each Tessera server type (`P2P`, `Q2T`, `ADMIN`, `THIRDPARTY`, `ENCLAVE`)?

If so, maybe remove this line and update "You can configure each.." with "You can optionally configure each..."

}
```

> where `truststore.jks` is a Java KeyStore format file containing the trusted certificates for the Tessera client (for

Reviewer 2: I think Besu uses "keystore".

example, the certificate of the CA used to create the InfluxDB certificate).

If securing the key store with a password, you must provide this password.
Passwords can be provided either in the configuration (for example `clientTrustStorePassword`) or as environment
variables (using `environmentVariablePrefix` and setting `<PREFIX>_TESSERA_CLIENT_TRUSTSTORE_PWD`).
The [TLS configuration](../Configure/TLS.md) documentation explains this in more detail.

As Tessera expects two-way TLS, a `.jks` file for the `clientKeyStore` must also be provided.
This isn't used so can simply be set as the trust store.

### Using Prometheus

> The [Prometheus documentation](https://prometheus.io/docs/introduction/overview/) provides all the information you need
> to get Prometheus setup and ready to integrate with Tessera.

Reviewer 2: ```suggestion
The [Prometheus documentation](https://prometheus.io/docs/introduction/overview/) provides information 
to set up Prometheus to integrate with Tessera.
```

The [Prometheus First Steps](https://prometheus.io/docs/introduction/first_steps/) is a good starting point.
A summary of the steps to store Tessera metrics in a Prometheus database are as follows:

1. Install Prometheus.
1. Create a `prometheus.yml` configuration file to provide Prometheus with the necessary information to pull metrics
   from Tessera.
   You can use [an example Prometheus configuration](https://github.com/ConsenSys/quorum-dev-quickstart/blob/master/files/goquorum/config/prometheus/prometheus.yml)
   with the [Quorum Developer Quickstart](../../Tutorials/Quorum-Dev-Quickstart.md).
1. Start Tessera.

>    As Tessera always exposes the `metrics` endpoint, no additional configuration of Tessera is required.

Reviewer 2: ```suggestion
   Tessera always exposes the `metrics` endpoint, so no additional configuration is required.
```

1. Start Prometheus:

    ```bash
    prometheus --config.file=prometheus.yml
    ```

1. To view data stored in the database, access the Prometheus UI (by default `localhost:9090`, this address can be
   changed in `prometheus.yml`) and use the [Prometheus Query Language](https://prometheus.io/docs/prometheus/latest/querying/basics/).

### Using Grafana

You can import a pre-built [GoQuorum Grafana dashboard](https://grafana.com/grafana/dashboards/14360) to visualize
your recorded GoQuorum network data.

## Monitoring Tessera

### Using Splunk

You can use Splunk to search, analyze, and monitor the logs of Tessera nodes.

> Consolidating the logs from multiple Tessera nodes in a network requires setting up Splunk and Splunk Universal Forwarders.

Reviewer 2: ```suggestion
To Consolidate logs from multiple Tessera nodes, set up Splunk and Splunk Universal Forwarders.
```

example, the certificate of the CA used to create the InfluxDB certificate).

If securing the key store with a password, you must provide this password.
Passwords can be provided either in the configuration (for example `clientTrustStorePassword`) or as environment
variables (using `environmentVariablePrefix` and setting `<PREFIX>_TESSERA_CLIENT_TRUSTSTORE_PWD`).
The [TLS configuration](../Configure/TLS.md) documentation explains this in more detail.

As Tessera expects two-way TLS, a `.jks` file for the `clientKeyStore` must also be provided.
This isn't used so can simply be set as the trust store.

### Using Prometheus

The [Prometheus documentation](https://prometheus.io/docs/introduction/overview/) provides all the information you need
to get Prometheus setup and ready to integrate with Tessera.
The [Prometheus First Steps](https://prometheus.io/docs/introduction/first_steps/) is a good starting point.
A summary of the steps to store Tessera metrics in a Prometheus database are as follows:

1. Install Prometheus.
1. Create a `prometheus.yml` configuration file to provide Prometheus with the necessary information to pull metrics
   from Tessera.

>    A simple Prometheus configuration for use with the [7nodes example network](../../../../Getting Started/Quorum-Examples) is:

Reviewer: I think we should replace all refs to the 7nodes example with the quickstart maybe? We can give them a direct link to example config too https://github.com/ConsenSys/quorum-dev-quickstart/blob/master/files/goquorum/config/prometheus/prometheus.yml and a grafana dashboard working
